src/controllers/bookController.ts

```diff
@@ -0,0 +1,128 @@
+import { NextFunction, Request, Response } from "express";
+import Book, { IBook } from "../models/Book";
+import { StatusCodes } from "http-status-codes";
+import { BadRequestError, NotFoundError } from "../errors";
+import mongoose from "mongoose";
+
+// Interface for query parameters
+interface IGetBooksQuery {
+  search?: string;
+  categories?: string;
+  sort?: "a-z" | "z-a" | "latest" | "oldest";
+}
+
+// Get all books
+const getAllBooks = async (
+  req: Request<{}, {}, {}, IGetBooksQuery>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, categories, sort } = req.query;
+  let query: any = {};
+
+  // Filter by categories
+  if (categories) {
+    const categoryArray = Array.isArray(categories)
+      ? categories
+      : categories.split(",");
+
+    query.categories = { $in: categoryArray };
+  }
+
+  // Search by title
+  if (search) {
+    query.title = { $regex: new RegExp(search, "i") }; // Case-insensitive search
+  }
+
+  try {
+    // Sort options
+    const sortOptions: any = {};
+    if (sort) {
+      if (sort === "a-z") sortOptions.title = 1;
```

Comment:
I think it would be more expressive to separate sorting by title and by publish date into separate parameters rather than creating this sort of "code" that you have to use to determine what field is being sorted in what way.

i.e.
```
sortTitle: string; // 'asc' or 'desc'
sortPublishDate: string // 'asc' or 'desc'
```

or even

```
sortBy: string; // 'title' or 'publishedDate`, reject anything else with an error
sortOrder: string; // 'asc' or 'desc'
```